### PR TITLE
Stop using CompareObjectHandles when not found

### LINKED ISSUE #2977: Windows support: replace the call to CompareObjectHandles function
As part of the checks performed to ensure that the caller process is still alive during workload attestation, we perform a system call to the CompareObjectHandles function, that compares two object handles to determine if they refer to the same underlying kernel object. This function requires at least Windows Server 2016.
Find a way to replace the call to the CompareObjectHandles function having the same functionality, so we can support earlier Windows versions.

### PR DESCRIPTION
**Pull request by MarcosDY**
The process ID is a value associated with the process object, and the process object remains as long as the process is still running or as long as somebody still has a handle to the process object. 
The Watcher verify process is still alive calling GetExitCodeProcess to retrieve the exit code, and an extra verification CompareObjectHandles was added, but it was introduced on Windows Server 2016, so for  older versions it is not available.
For those scenarios we rely on GetExitCodeProcess is enough.

**Which issue this PR fixes**
fixes #2977

### CODE REVIEW

**Comment by azdagron** [comment]
I think this sounds ok? We have good evidence that Windows will not reuse a process ID as long as there remains an open handle to the process. Since we open the handle right after accept, and the verify the handle is still valid via GetExitCodeProcess,  seems like the CompareObjectHandles call doesn't really add to the security.
@amartinezfayo @evan2645 @rturner3 ? Thoughts?

**Comment by evan2645** [comment]
👍 I think it's generally ok for now

What I have learned from working on peertracker over the years is that there are almost always surprises to be had, and practically always, they are detrimental to peertracker security. Defense in depth in this part of the codebase is really important, and the more mitigating measures we can pack in the better... defense in depth in Linux peertracker has in the past saved us.

@anvega we are still oscillating on this critical windows workload attestation code .. I think you reached out in TAG Security to see if there was any movement related to https://github.com/cncf/tag-security/issues/853, any news on that front?

This PR looks good to me.

File: pkg/common/peertracker/tracker_windows.go [hunk 1]
```diff
@@ -174,6 +179,9 @@ type systemCaller interface {
 	// GetExitCodeProcess retrieves the termination status of the
 	// specified process handle.
 	GetExitCodeProcess(windows.Handle, *uint32) error
+
+	// IsCompareObjectHandlesFound verifies if CompareObjectHandles is found on Windows instance
```
Comment (amartinezfayo) [inline pkg/common/peertracker/tracker_windows.go:183]: nit:
```suggestion
	// IsCompareObjectHandlesFound returns true if the CompareObjectHandles function could be found in this Windows instance
```

**Review by amartinezfayo — COMMENTED** [review]
I think that this approach looks good and I'm OK with not having the extra security measure of calling CompareObjectHandles when it's not available. The main security measure that we have is 1) open the handle of the process as soon as we can to avoid the PID reuse and 2) check for the exit code of the handle to ensure that the process is still active.
It may worth investigating what CompareObjectHandles may be doing internally. It's a possibility that we can call a set of APIs and achieve the same/similar result as calling this single function.

**Comment by evan2645** [comment]
> The main security measure that we have is 1) open the handle of the process as soon as we can to avoid the PID reuse and 2) check for the exit code of the handle to ensure that the process is still active.
> It may worth investigating what CompareObjectHandles may be doing internally.

Just wanna say again I think this PR is fine, but in the interest of sharing experiences: the last time I recall we got surprised by peertracker .. the Linux peertracker was using some similar handle tricks against procfs, and we were relying on the ability to read something off that handle as telling whether or not the file/process still existed.

Turns out that doing this would normally work, except we were using some stdlib stuff for handling the filesystem interactions and stdlib was going some proactive caching for perform reasons ... And busting the whole mitigation. The fix was to move to a lower level of interaction.

Luckily, we still had UID and starttime checks which together were strong enough to mitigate the file handle breakage.

Credit and healthy kudos to @azdagron on picking that one apart 🤗

**Comment by amartinezfayo** [comment]
> Luckily, we still had UID and starttime checks which together were strong enough to mitigate the file handle breakage.

We also explored those mitigations in Windows, and with the presence of the CompareObjectHandles function we decided that comparing the object handles would be a more robust check.

Thinking that the mitigations 1) and 2) could be bypassed due to circumstances that we are not able to recognize at this point in time is a fair point. I wouldn't be opposed to have the UID and starttime checks when CompareObjectHandles is not available, if we all feel more comfortable with that. Here is an example implementation:
https://github.com/spiffe/spire/pull/2695/commits/ceb5a3b57b8444a20f937fff76f70d41cd20c212#diff-532ab68dea9dbe10a91514f4052bc0f1a5b7ecaef4b51235e22faca439c78dacR118

**Comment by MarcosDY** [comment]
about time, I'm worried about clock changes... But I already have code in this branch history to verify CreationTime https://github.com/spiffe/spire/pull/2999/commits/7df17e43749f51fb1a6d427d06fa88874bfe724e

**Review by azdagron — APPROVED** [review]
We talked as maintainers and agreed this is an ok step. We will open an issue to explore additional mitigations.

**Pull request merged**